File: core/src/com/bombbird/terminalcontrol/screens/BasicScreen.java
package com.bombbird.terminalcontrol.screens;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.utils.viewport.FitViewport;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.bombbird.terminalcontrol.TerminalControl;

public class BasicScreen implements Screen {
    public final TerminalControl game;
    public final Stage stage;
    public final OrthographicCamera camera;
    public final Viewport viewport;

    public BasicScreen(final TerminalControl game, int width, int height) {
        this.game = game;

        //Set camera params
        camera = new OrthographicCamera();
        camera.setToOrtho(false, width, height);
        viewport = new FitViewport(TerminalControl.WIDTH, TerminalControl.HEIGHT, camera);
        viewport.apply();

        //Set stage params
        stage = new Stage(new FitViewport(width, height));
        stage.getViewport().update(TerminalControl.WIDTH, TerminalControl.HEIGHT, true);
        Gdx.input.setInputProcessor(stage);
    }

    /** Implements show method of screen, should be overridden by individual classes */
    @Override
    public void show() {
        //No default implementation
    }

    /** Main rendering method for rendering to spriteBatch */
    @Override
    public void render(float delta) {
        Gdx.gl.glClearColor(0, 0, 0, 0);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT | GL20.GL_DEPTH_BUFFER_BIT | (Gdx.graphics.getBufferFormat().coverageSampling?GL20.GL_COVERAGE_BUFFER_BIT_NV:0));

        camera.update();

        game.batch.setProjectionMatrix(camera.combined);
        stage.act(delta);
        game.batch.begin();
        boolean success = false;
        while (!success) {
            try {
                stage.draw();
                game.batch.end();
                success = true;
            } catch (IndexOutOfBoundsException e) {
                Gdx.app.log(this.getClass().getName(), "stage.draw() render error");
                stage.getBatch().end();
                e.printStackTrace();
            }
        }
    }

    /** Implements resize method of screen, adjusts camera & viewport properties after resize for better UI */
    @Override
    public void resize(int width, int height) {
        viewport.update(width, height, true);
        stage.getViewport().update(width, height, true);
        camera.position.set(camera.viewportWidth / 2, camera.viewportHeight / 2, 0);
    }

    /** Implements pause method of screen */
    @Override
    public void pause() {
        //No default implementation
    }

    /** Implements resume method of screen */
    @Override
    public void resume() {
        //No default implementation
    }

    /** Implements hide method of screen */
    @Override
    public void hide() {
        dispose();
    }

    /** Implements dispose method of screen */
    @Override
    public void dispose() {
        stage.clear();
        stage.dispose();
    }
}

File: core/src/com/bombbird/terminalcontrol/screens/BasicScreen.kt
package com.bombbird.terminalcontrol.screens

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Screen
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.utils.viewport.FitViewport
import com.badlogic.gdx.utils.viewport.Viewport
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.utilities.SafeStage

open class BasicScreen(val game: TerminalControl, width: Int, height: Int) : Screen {
    val stage: SafeStage
    val camera: OrthographicCamera = OrthographicCamera()
    private val viewport: Viewport

    init {
        //Set camera params
        camera.setToOrtho(false, width.toFloat(), height.toFloat())
        viewport = FitViewport(TerminalControl.WIDTH.toFloat(), TerminalControl.HEIGHT.toFloat(), camera)
        viewport.apply()

        //Set stage params
        stage = SafeStage(FitViewport(width.toFloat(), height.toFloat()), game.batch)
        stage.viewport.update(TerminalControl.WIDTH, TerminalControl.HEIGHT, true)
        Gdx.input.inputProcessor = stage
    }

    /** Implements show method of screen, should be overridden by individual classes  */
    override fun show() {
        //No default implementation
    }

    /** Main rendering method for rendering to spriteBatch  */
    override fun render(delta: Float) {
        Gdx.gl.glClearColor(0f, 0f, 0f, 0f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT or GL20.GL_DEPTH_BUFFER_BIT or if (Gdx.graphics.bufferFormat.coverageSampling) GL20.GL_COVERAGE_BUFFER_BIT_NV else 0)
        camera.update()
        game.batch.projectionMatrix = camera.combined
        stage.act(delta)
        stage.draw()
    }

    /** Implements resize method of screen, adjusts camera & viewport properties after resize for better UI  */
    override fun resize(width: Int, height: Int) {
        viewport.update(width, height, true)
        stage.viewport.update(width, height, true)
        camera.position[camera.viewportWidth / 2, camera.viewportHeight / 2] = 0f
    }

    /** Implements pause method of screen  */
    override fun pause() {
        //No default implementation
    }

    /** Implements resume method of screen  */
    override fun resume() {
        //No default implementation
    }

    /** Implements hide method of screen  */
    override fun hide() {
        dispose()
    }

    /** Implements dispose method of screen  */
    override fun dispose() {
        stage.clear()
        stage.dispose()
    }
}